Doc/library/importlib.rst

```diff
@@ -1284,66 +1307,101 @@ find and load modules.
+   non-package modules run from standard input.
 
    .. attribute:: loader
 
    (``__loader__``)
 
-   The loader to use for loading.  For namespace packages this should be
-   set to ``None``.
+   The loader to use for importing the module.
+   It is not set for non-package modules run from the file system and
+   non-package modules run from standard input.
 
    .. attribute:: origin
 
    (``__file__``)
 
-   Name of the place from which the module is loaded, e.g. "builtin" for
-   built-in modules and the filename for modules loaded from source.
-   Normally "origin" should be set, but it may be ``None`` (the default)
-   which indicates it is unspecified (e.g. for namespace packages).
+   The place from which the module's data was imported.  Its value and meaning
+   is up to the finder.
+   It is the relative path for packages run from the file system and the
+   absolute path for imported modules and modules run from the module
+   namespace.
+   It is ``None`` for imported namespace packages, ``'built-in'`` for imported
+   built-in modules and ``'frozen'`` for imported frozen modules.
+   It is not set for non-package modules run from the file system and
+   non-package modules run from standard input.
 
    .. attribute:: submodule_search_locations
 
    (``__path__``)
 
-   List of strings for where to find submodules, if a package (``None``
-   otherwise).
+   The (possibly empty) iterable of all locations to search (i.e. the search
+   path) when looking for the package's submodules.  Order is significant.
+   It is ``None`` for packages run from the file system, imported non-package
+   modules and modules run from the module namespace.
+   It is not set for non-package modules run from the file system and
+   non-package modules run from standard input.
 
    .. attribute:: loader_state
 
-   Container of extra module-specific data for use during loading (or
-   ``None``).
+   The finder may set this to an object containing additional data for the
+   loader to use when loading the module.  Otherwise it will be ``None``.
 
    .. attribute:: cached
 
    (``__cached__``)
 
-   String for where the compiled module should be stored (or ``None``).
+   The location of the module's cached data.  Some loaders support optimization
+   through caching the bytecode of the module.  ``cached`` is how the finder
+   tells the loader where to look for that cached bytecode, or where to store
+   it.
+   It is the relative path for packages run from the file system and the
+   absolute path for imported modules and modules run from the module
+   namespace.
+   It is ``None`` for imported namespace packages, imported built-in modules
+   and imported frozen modules.
+   It is not set for non-package modules run from the file system and
+   non-package modules run from standard input.
 
    .. attribute:: parent
 
    (``__package__``)
 
-   (Read-only) Fully-qualified name of the package to which the module
```

Comment:
As with `__package__`:

```
   (Read-only) The fully-qualified name of the package under which the
   module should be loaded as a submodule.
   For packages, the same as ``name``.
   For top-level modules, the empty string.
   For all other modules, the parent package's ``name``.
```

That "None" part was definitely a mistake.

Reply:
Alright.